### ISSUE TITLE
Roslyn analyzer support

### ISSUE THREAD

**Issue by KuraiAndras** [xoofx/UnityNuGet#66]
It would be nice if we could use this with Roslyn Analyzers. If I have some free time I might look into it.

**Comment by bdovaz** [comment]
Done! #67

**Comment by bdovaz** [comment]
@KuraiAndras it's already merged and when @xoofx releases this version in Azure you can close this issue

**Comment by KuraiAndras** [comment]
@bdovaz @xoofx something is not right with the implementation.
Unity picks up the analyzers, but the Visual Studio plugin does not, and they don't show up as analyzers inside Visual Studio. I noticed this when trying out source generators (I tried the Microsoft.Toolkit.MVVM package). They also work with your new feature. This means that you can use generated code, but Visual Studio will not detect the generated parts, but Unity detects it and compiles it successfully (even produces a working build)

The problem is (I think), that right now analyzers are disabled on all platforms, but they should not be. According to the [documentation](https://docs.unity3d.com/2021.2/Documentation/Manual/roslyn-analyzers.html) only the Editor and the Standalone platforms should be disabled.

I got this fixed by manually placing the analyzer DLL inside the project, and removed only the Editor and the Standalone platform.

I would like to fix it but I don't know how should this be handled (I presume we would have to hard code the supported platforms and update it with each unity release?).

<details>
  <summary>Working meta file</summary>

```yml
fileFormatVersion: 2
guid: 0faed146a08939941b39db6a69285ea5
labels:
- RoslynAnalyzer
PluginImporter:
  externalObjects: {}
  serializedVersion: 2
  iconMap: {}
  executionOrder: {}
  defineConstraints: []
  isPreloaded: 0
  isOverridable: 0
  isExplicitlyReferenced: 0
  validateReferences: 1
  platformData:
  - first:
      : Any
    second:
      enabled: 0
      settings:
        Exclude Android: 0
        Exclude Editor: 1
        Exclude Linux64: 1
        Exclude OSXUniversal: 1
        Exclude WebGL: 0
        Exclude Win: 1
        Exclude Win64: 1
        Exclude WindowsStoreApps: 0
        Exclude iOS: 0
  - first:
      Android: Android
    second:
      enabled: 1
      settings:
        CPU: ARMv7
  - first:
      Any: 
    second:
      enabled: 0
      settings: {}
  - first:
      Editor: Editor
    second:
      enabled: 0
      settings:
        CPU: AnyCPU
        DefaultValueInitialized: true
        OS: AnyOS
  - first:
      Standalone: Linux64
    second:
      enabled: 0
      settings:
        CPU: None
  - first:
      Standalone: OSXUniversal
    second:
      enabled: 0
      settings:
        CPU: None
  - first:
      Standalone: Win
    second:
      enabled: 0
      settings:
        CPU: None
  - first:
      Standalone: Win64
    second:
      enabled: 0
      settings:
        CPU: None
  - first:
      WebGL: WebGL
    second:
      enabled: 1
      settings: {}
  - first:
      Windows Store Apps: WindowsStoreApps
    second:
      enabled: 1
      settings:
        CPU: AnyCPU
        DontProcess: false
        PlaceholderPath: 
        SDK: AnySDK
        ScriptingBackend: AnyScriptingBackend
  - first:
      iPhone: iOS
    second:
      enabled: 1
      settings:
        AddToEmbeddedBinaries: false
        CPU: AnyCPU
        CompileFlags: 
        FrameworkDependencies: 
  userData: 
  assetBundleName: 
  assetBundleVariant:   
```
</details>

<details>
  <summary>Current meta file</summary>

```yml
fileFormatVersion: 2
guid: 5d1110779f690c57bedab8618120c663
labels:
  - RoslynAnalyzer
PluginImporter:
  externalObjects: {}
  serializedVersion: 2
  iconMap: {}
  executionOrder: {}
  isPreloaded: 0
  isOverridable: 0
  isExplicitlyReferenced: 0
  validateReferences: 1
  platformData:
  - first:
      Any:
    second:
      enabled: 0
      settings: {}
  - first:
      Editor: Editor
    second:
      enabled: 0
      settings:
        DefaultValueInitialized: true
  - first:
      Windows Store Apps: WindowsStoreApps
    second:
      enabled: 0
      settings:
        CPU: AnyCPU
  userData: 
  assetBundleName: 
  assetBundleVariant: 
```

</details>

You can see a sample at my branch https://github.com/KuraiAndras/UnityNuGet-1/tree/add-serilog-extensions.
At the commits:
- [Here](https://github.com/KuraiAndras/UnityNuGet-1/tree/d9698f03ac7d118cfa842bca79c9d7040775d177) is the working state
- [Here](https://github.com/KuraiAndras/UnityNuGet-1/tree/fa2e78a5531cfe8bcec876f13db30f9eb97db929) is the current solution

To reproduce it locally add the following packages to the registry:
```json
  "Microsoft.Toolkit.Mvvm":{
    "listed": true,
    "version": "7.0.0",
    "analyzer": true
  },
  "System.ComponentModel.Annotations":{
    "listed": true,
    "version": "4.4.0"
  },
```

**Comment by KuraiAndras** [comment]
Okay, so I was wrong, the problem is not with UnityNuGet, but with the Visual Studio extension for Unity. It generates the wrong paths for analyzers when they are located in the Package cache.
```xml
<ItemGroup>
    <!-- Original -->
    <Analyzer Include="C:\Work\UnityNuGet-1\examples\unity\UnitySample\Packages\org.nuget.microsoft.toolkit.mvvm\dotnet\cs\Microsoft.Toolkit.Mvvm.SourceGenerators.dll" />
    <!-- Fixed -->
    <Analyzer Include="C:\Work\UnityNuGet-1\examples\unity\UnitySample\Library\PackageCache\org.nuget.microsoft.toolkit.mvvm@7.1.1\dotnet\cs\Microsoft.Toolkit.Mvvm.SourceGenerators.dll" />
</ItemGroup>
```

The good new is that both analyzers and source generators work now. The bad news is that the VS integration is buggy.

**Comment by KuraiAndras** [comment]
So the problem is a bit strange. It seems like is Unity doing some symlink magic (maybe?) where files located in the packages cache are also available from the Packages folder, but Visual Studio does not see them. Here is an asset post process which fixes it (Path.GetFullPath will return the actual location of the file for some reason)

```csharp
#if UNITY_EDITOR
using System.IO;
using System.Linq;
using System.Xml.Linq;
using UnityEditor;

public sealed class AnalyzerFixer : AssetPostprocessor
{
    public static string OnGeneratedCSProject(string path, string content)
    {
        if (Path.GetExtension(path) != ".csproj") return content;

        var csproj = XDocument.Parse(content, LoadOptions.PreserveWhitespace);

        FixAnalyzers(csproj);

        using var writer = new StringWriter();
        csproj.Save(writer);

        return writer.ToString();
    }

    private static void FixAnalyzers(XDocument csproj)
    {
        foreach (var analyzer in csproj
            .Elements()
            .SingleOrDefault(e => e.Name.LocalName == "Project")
            ?.Elements()
            .Where(e => e.Name.LocalName == "ItemGroup")
            .SelectMany(itemGroup => itemGroup.Elements().Where(e => e.Name.LocalName == "Analyzer")))
        {
            var includeAttribute = analyzer.Attribute("Include");
            var include = Path.GetFullPath(includeAttribute.Value);

            includeAttribute.SetValue(include);
        }
    }
}
#endif

```

**Comment by bdovaz** [comment]
Ping @jbevain, @sailro

**Comment by sailro** [comment]
Thank you for the ping we'll have a look. I think we are only passing the path to analyzer given by Unity when generating csproj files. The only thing we do on our side is to automatically add our own Analyzer.

Are you on the latest VS Editor package v1.0.12 ?

**Comment by KuraiAndras** [comment]
Yes the latest 2.0.12

**Comment by KuraiAndras** [comment]
Also I did not try this with the Rider or the VSCode package, they also might be affected

**Comment by KuraiAndras** [comment]
I tried it out with the VSCode package, and that has the same problem, so I suppose the rider package also has this issue

**Comment by bdovaz** [comment]
Well, if it fails in VS and VSCode it looks more like it's an internal problem of Unity and not of VS or VSCode Unity packages.

@sailro I don't know if you have been able to spend some time investigating the problem to confirm it.

If it is a Unity problem @sailro @xoofx can you help in any way to get Unity to investigate this problem? Thanks.

**Comment by sailro** [comment]
I'm having a look

**Comment by sailro** [comment]
Okay we'll fix on our side.

Indeed it is working in Unity because of the `PackageCache` transparent redirection. But we already handle this for other cases. So we'll do the same for this. The fix will be available with the next 1.0.13 package. 

As a workaround for now, you can use the AssetPostprocessor given by @KuraiAndras, because simply using `Path.GetFullPath` will properly resolve the proper filename.

Thanks!

**Comment by xoofx** [comment]
Pushing the new server 0.13.0, thanks for the idea @KuraiAndras, the contribution @bdovaz, and the fix in VS @sailro!
